### ISSUE TITLE
ui5-automated testing recording

### ISSUE THREAD

**Issue by timostark** [SAP/ui5-uiveri5#7]
Hi All,

First thanks for that repository - E2E testing (support/examples) was one of the missing features of the SAPUI5 framework (even though i personally like testcafe more than protractor).

What I'd like to understand: Is this library of strategic meaning for sap - in sense that it will be used as common end-to-end testing framework library for sapui5? Is the tool used SAP internally?

_Background_: I am writing a ui5-test recorder (chrome / firefox plugin) at the moment for testcafe (testcafe plugin: https://github.com/timostark/ui5TestcafeSelector), which seems to have a very similar approach to ensure test-stability compared to what you are using (don't only select controls by a ID, especially for lists, but instead try to find the best matching combination of attributes possible - see example below). An adaption for protractor / uiveri5 doesn't seem to be too difficult.

Therefore two questions:
1. Is the project accepting pull requests, to extend the selector capabilities (e.g. select by aggregations, associations, or binding paths)?
2. Is the repository used sap internal as strategic tool for sapui5 e2e testing?

![sapui5_testing](https://user-images.githubusercontent.com/19237828/51185237-dceb8000-18d6-11e9-8679-747e9c279eaa.gif)

Thanks a lot for an honest response :-)

Regards,
Timo

**Comment by maximnaidenov** [comment]
Hi Timo,

This tool was open-sourced just few days ago but is used inside SAP for 3 years now. Unfortunately this is not visible from the project history as one of the requirements was to squash history to single initial commit.
I can not state this "the official E2E tool" as SAP is pretty large organisation and there are a lot of teams with different requirements, usecases and tool preferences. 
What I can state is that we, the team behind this tool is the UI5 development e.g. part of the organisation that develops SAPUI5/OpenUI5. And we are the same team that develops and maintains OPA.
So we are the nearest to "official" that you can find, IMHO. Also, our work is OSS so you have the chance to fork and develop whatever you like on top, actually we have several teams inside SAP do it.
As for selectors you mention, they are already implemented - everything you have in OPA as matchers is available declaratively as control locators. Please check the [docs](https://github.com/SAP/ui5-uiveri5/blob/master/docs/usage/locators.md) for details.

BTW.
Some recording/snippet generation functionality is in our plans for this year so I would be interested to see your work and possibly inspire.. We even have a RecordReplay API already in place, please check the [demokit API docs](https://openui5nightly.hana.ondemand.com/#/api/sap.ui.test.RecordReplay).

Maxim

**Comment by timostark** [comment]
Hi Maxim,

Thanks - That is more than enough of an answer (especially, that behind that project is the _ui5 development team_). I will adapt the recording on my side. Thanks also for the link to the documentation. You are right - I assume the OPA5 matchers are sufficient.

Two additional question/remarks: Protractor is imo based on selenium. I am personally not the biggest fan of selenium (without full control, it is mostly difficult for the framework to perfectly determine wait-times for assertions and selectors; so sometimes in case of "hanging" XHR requests or slow animation requests, you are getting an false-positive alert..additionally mobile testing is certainly difficult). Have you evaluated a switch to testcafe or cypress?

2. Yes I saw the record replay on openui5 nightly (and was slightly annoyed, that I saw it after hacking around to simulate press & click events correctly =) ).  Are you planning to use that API for *running* the tests, or just for recording previews (of a potential recorder, created that year)?

For "inspiration" you can have a look on https://chrome.google.com/webstore/detail/hcpkckcanianjcbiigbklddcpfiljmhj/publish-accepted?authuser=0&hl=de - Don't expect sth. stable though. It is for my personal fun ;-)

Regards,
Timo

**Comment by maximnaidenov** [comment]
Hi Timo,

uiveri5 is based on protractor (for now, actually on ideas from protractor) with the main one being the automatic synchronization with UI5 runtime. So there is no need to determine wait-time, the .click() will happen only after UI5 is fully rendered. Similarly, expect() will happen only when the value is ready.
Of course, there are timeouts guarding hangups but the tests does not need single wait() or sleep() and works as fast as your browser/backend can render ui5.
The implementation is similar in concept with zonejs. In this respect, we are much ahead of cypress as we also synchronise with deferred work (settimeout), promises, rendering pipeline, some specific controls like NavContainer, etc.
Testing on mobile devices works perfectly fine, this was one of the original usecases.
RecordReplay API was implemented as contribution to current and upcoming SAP products. You can use it for recording only and then implement a replay with selenium (getDOMElementForControlSelector) or you can use it for in-browser replay (interactWithControl), similarly to OPA5. Both usecases are supported.

Regards,
Maxim

**Comment by timostark** [comment]
Hi Maxim,

Got it. Thanks for the detailed explanation. You can get inspired at https://chrome.google.com/webstore/detail/ui5-test-recorder/hcpkckcanianjcbiigbklddcpfiljmhj (private package for now).

Test e.g. against https://openui5.hana.ondemand.com/test-resources/sap/m/demokit/cart/webapp/index.html# , clicking the addon-menu and press "new test".

Greets,
Timo

**Comment by timostark** [comment]
OK again, thanks for the stable framework. I noticed a few strange behaviors, when testing (e.g. you have to set the interaction to "root", when making an AggregationEmpty assertion on a List).
I updated the plugin, to have a selector and a recorder for ui5-ui5veri5 (1000 of bugs).
If you want you can have a look on it of course.

Remarks:
- As far as I see, it is not possible to set the strictSSL=truefor the request npm module. While it anyways it is not a good idea to set it to true, it might provoke some defects for people working behind a corporate proxy
- It is required to set interaction to "root" for aggregationEmpty bindings, to avoid "no-dom-existing" errors from the action-provider. Is that intended behavior? If no I can of course raise a separate defect - just let me know.

Sample:
test.spec.js
```
describe('test', function () {
    it('Test 0', function () {
        element(by.control({ id: /homeView--searchField/, interaction: { idSuffix: "I" } })).sendKeys('NOT_EXISTING');
        expect(element.all(by.control({ controlType: "sap.m.List", id: /productList/, aggregationEmpty: { name: "items" }, interaction: "root" })).count()).toBeGreaterThan(0);
    });
});
```

conf.js
```
exports.config = {
    profile: 'integration',
    baseUrl: 'https://openui5nightly.hana.ondemand.com/test-resources/sap/m/demokit/cart/webapp/index.html'
};
```

**Comment by maximnaidenov** [comment]
Hi Timo,

I do not get what you mean with the strictSSL, isn't that a setting for npm ?. We publish on npm in a pretty standart way so I don't think this has any relation to our project ?.
As for the ancestor and interaction type, some remarks:
* having both id and control type is excessive, id take precedence. Same in OPA
* the aggregationEmpty, etc is supported in uiveri5 for consistency but do not have so much sense and we always synchronize before interactions. So you could simply get the table and assert the rows or get the rows and assert their count, etc. The aggregation matchers are an artifact from the past of OPA, before having the synchronisation ( autoWait: true) and we keep them for backward compatability
* interaction type should be specified in very specific usecases where you have several "actions" on a control and you want to excersise non-default one.  Like in a searchfield, etc. Otherwise, all controls should work without interaction type. Can you please open another issue and copy the example and we will have a look.

Maxim
